### PR TITLE
Start Refining Errors

### PR DESCRIPTION
**Pull request by kazk**
I wanted to add "Refine Errors" (opened #688) to our roadmap, so I tried a bit to see how we can tackle this. Any feedback is appreciated.

### Remove `impl From<T> for kube::Error`

Require explicitly mapping errors.
[This commit](https://github.com/kube-rs/kube-rs/commit/f595f9bd960abf959206055e378a418ea49dd86b) shows how much we need to mentally process while reading the code (maintenance cost), and how badly grouped they are (usability issue).
Removing `impl From` forces us to attach context, and this is a good thing. It's verbose right now, but that's because we need to break them down.

As an example, I got rid of `kube_core::Error`. Parsing group version no longer has an error variant `HttpError`.

### `thiserror` vs `snafu`

As I mentioned in [`thiserror` vs `snafu`](https://github.com/kube-rs/kube-rs/discussions/453#discussioncomment-451968) discussion, the huge `kube::Error` is a problem I'd like us to start fixing. I've been using both `thiserror` and `snafu`, and I think using `thiserror` without `#[from]`, and gradually breaking errors down is probably the best approach for `kube`. While I think `snafu` help us _initially to think_ of errors in a more scalable way, it's not necessary to implement the pattern. Also, we probably can't do SNAFU way until we finish breaking them down cleanly anyway, and the features to prevent misuse will get in the way (e.g., private context selectors by default). So, I don't think it's worth some downsides.

#### Stability

`thiserror` is stable (v1). `snafu` is still evolving. 0.7 will add `Snafu` suffix to generated context selectors. Tuple variants are [not well-supported](https://github.com/shepmaster/snafu/issues/61)? (I haven't used tuple variants with `snafu`, so I don't know the details).

#### Popularity

`thiserror` v1 is used by [9,491](https://lib.rs/crates/thiserror/rev) public crates. `snafu` 0.6 is used by [270](https://lib.rs/crates/snafu/rev). `snafu` is _extra_ dependency for many.

#### Features

`snafu` provides some syntactic sugars, but I don't think it adds that much value. Avoiding `#[from]` and using the usual `Result::map_err` feels similar enough, especially as we break the errors down.

```rust
let mut file = File::open(path).context(OpenFileSnafu { path })?;
let mut file = File::open(path).map_err(|e| Error::OpenFile(e, path.to_owned()))?;
let mut file = File::open(path).map_err(|source| Error::OpenFile { source, path: path.into() })?;

action().context(ActionSnafu)?;
action().map_err(Error::Action)?;
```

One feature `snafu` optionally provides that `thiserror` doesn't, is backtrace in stable, but this requires the end users to depend on `snafu` and enable the `backtrace` feature. There are error reporter crates to do this until backtrace become stable.

Unless I missed something significant, I think we should go with `thiserror` without `#[from]`.

### CODE REVIEW

**Comment by teozkr** [comment]
Big :+1: on the overall idea.

> I've been using both thiserror and snafu, and I think using thiserror without #[from], and gradually breaking errors down is probably the best approach for kube.

That makes sense to me.

> While I think snafu help us initially to think of errors in a more scalable way, it's not necessary to implement the pattern.

Definitely, it's all just syntax sugar in the end.

> Also, we probably can't do SNAFU way until we finish breaking them down cleanly anyway, and the features to prevent misuse will get in the way (e.g., private context selectors by default). So, I don't think it's worth some downsides.

While it'd be nice to only depend on one of them in the end, we don't need to go all-in from the start. Neither of them are visible in the public API, and they interoperate just fine (aside from SNAFU's backtraces).

> thiserror is stable (v1). snafu is still evolving. 0.7 will add Snafu suffix to generated context selectors.

I'm not looking forward to that upgrade, but there's also no huge rush. We're not holding anyone else back from upgrading since we're only exposing the `std::error::Error` instance anyway.

> Tuple variants are not well-supported? (I haven't used tuple variants with snafu, so I don't know the details).

To be honest I kind of dislike tuple variants anyway, since they become quite messy as soon as you want to add any extra context to the errors, as demonstrated by your file example. I'd feel much more comfortable pattern-matching `Error::OpenFile { source: io::Error, path: PathBuf }` than `Error::OpenFile(io::Error, Path)` (and this is a relatively tame example since all context has types that more or less indicate what everything is..).

**Comment by teozkr** [comment]
To be clear: even if I might personally prefer going full SNAFU, (IMO) modularized thiserror is still a vast improvement over what we currently have.

**Comment by clux** [comment]
Yeah, I like this a lot as well. Gets us going on refining.

> `thiserror` vs `snafu`

This justification is good. The popularity of [`thiserror`](https://lib.rs/crates/thiserror/rev) has exploded compared to [`snafu`](https://lib.rs/crates/snafu/rev) which is declining (and we are the top listed depender). Thiserror They do also have `dtolnay` behind it as a maintainer who is making frequent non-breaking patch releases (.30 now).

Regardless of how pointless a popularity contest may seem, pretty much everyone knows how to wrangle `thiserror` and that is also nice from a fresh contributor POV.

**Comment by kazk** [comment]
Cool, I'll open a tracking issue with a summary and task list.

> To be honest I kind of dislike tuple variants anyway, since they become quite messy as soon as you want to add any extra context to the errors, as demonstrated by your file example. I'd feel much more comfortable pattern-matching `Error::OpenFile { source: io::Error, path: PathBuf }` than `Error::OpenFile(io::Error, Path)` (and this is a relatively tame example since all context has types that more or less indicate what everything is..).

Yeah, we should avoid tuple if it makes it difficult for users. We should decide what's best for each error, and `thiserror` doesn't limit us.

File: kube-runtime/src/controller/mod.rs
Comment (kazk) [inline kube-runtime/src/controller/mod.rs:48]: I think tuples are fine here, but we can do struct variants if that's better. 

Added `Display` to match `snafu`'s default (variant name, then inner error if there's a source).

File: kube-runtime/src/wait.rs
Comment (kazk) [inline kube-runtime/src/wait.rs:13]: Alternatively, we can do `pub struct ProbeError(#[source] watcher::Error)`.

File: kube-client/src/error.rs
Comment (clux) [inline kube-client/src/error.rs:24]: Does this not give us a source error with the source's error msg in the display? Should we not be removing `{0}`'s here as per recommendations discussed in https://github.com/kube-rs/kube-rs/pull/515 ?

File: kube-runtime/src/controller/mod.rs
Comment (clux) [reply kube-runtime/src/controller/mod.rs:48]: Imo this is fine. Tuples matches the approach we follow elsewhere.

File: examples/configmapgen_controller.rs
Comment (clux) [inline examples/configmapgen_controller.rs:24]: for consistency, this should probably be a tuple

**Review by clux — COMMENTED** [review]
This looks great to me! Generally just minor nits. But but have one question left on our approach w.r.t. https://github.com/kube-rs/kube-rs/pull/515 .

File: kube-client/src/error.rs
Comment (clux) [reply kube-client/src/error.rs:24]: Ah, I see that's out of scope for this PR but listed in the overall goals. :+1:

Comment (kazk) [reply kube-client/src/error.rs:24]: Yeah, I added a task for that in #688. All of our errors with source does this, so I think we should change all of them in a separate PR.

**Review by clux — APPROVED** [review]

**Pull request merged**
